Improve xccdf_results_mapper when converting XCCDF Results to HDF Results

candrews opened the pull request:
* Support XCCDF generated using ComplianceAsCode: https://github.com/ComplianceAsCode/content/
* Support nested XCCDF groups
* Set `code_desc` to an appropriate value (not empty string)
* Refactor to remove global variables

Bug: https://github.com/mitre/saf/issues/1163
Fixes: https://github.com/mitre/heimdall2/issues/4194

aaronlippold commented:
Hi the team @Amndeep7 @em-c-rod @wdower @Hookwitz @georgedias are looking over this and seeing if some of our other PR work can be rolled into this as well. 

A couple questions:
- Given some of the changes that were make, would we be able to make some smaller unit tests that cover actual function @em-c-rod @Amndeep7 @candrews 
- Should we add test data for other XCCDF profiles -windows, Ubuntu, DB, IIS, etc or would a subset of XCCDF content element likely be sufficient
- Have we seen any corner cases - nesting, tags inside tags, etc - that we should make specific tests for?

aaronlippold commented:
@candrews looks like SonarCloud is also unhappy about something - which we may want to poke at as well as doing a rebase from master/main

aaronlippold commented:
Question: Is any of the work here also useful or overtaken by this PR? https://github.com/mitre/heimdall2/pull/2958

candrews commented:
> looks like SonarCloud is also unhappy about something - which we may want to poke at as well as doing a rebase from master/main

I fixed the SonarCloud findings. The only finding now is in the test for a comment (the same comment is in every other test, so I think it should be in this one as well) and I believe that finding should be suppressed.

candrews commented:
> Question: Is any of the work here also useful or overtaken by this PR? #2958

This PR completely supersedes #2958 - it does everything that PR does and more.

candrews commented:
Can I do something to help this along?

I'd really like to have these issues fixed soon as it's blocking progress on a project which is import to my corporate overlords.

Amndeep7 commented:
Sorry for the delay @candrews - could you address the findings for the HDF-Converters tests?  It seems like some of the other XML mappers such as the Veracode are now having issues.

Amndeep7 commented:
Another point that I wanted to bring up was that we were intending on eventually rewriting this mapper to use jsonix to be able to match against the XCCDF schema properly instead of using the fast-xml-parser which has issues with (for example) determining the difference between singular nodes and arrays of length one.  This isn't a request for you to rework this entire PR to use that other library, just a heads up about part of the delay in dealing with this mapper since we're still exploring how to use jsonix and write a mapper properly with it in our checklist mapper work (https://github.com/mitre/heimdall2/pull/3859).

Amndeep7 commented on libs/hdf-converters/src/xccdf-results-mapper.ts:
```diff
@@ -23,79 +23,70 @@ const IMPACT_MAPPING: Map<string, number> = new Map([
+  benchmark: Record<string, unknown>
+): Record<string, unknown> | undefined {
+  for (const pathRuleResult of [
+    'cdf:TestResult.cdf:rule-result',
```
After we'd written this, we saw that there might be ways to make this easier.

https://github.com/mitre/heimdall2/issues/3084

The blocker for this is updating the version of the fast-xml-parser, which I could look into hopefully soon.

Amndeep7 replied:
That got merged

candrews replied:
I've updated this MR and this change certainly does make the task much easier and less error prone!

Amndeep7 commented:
Still need to finish reading/reviewing the code, which I'll try to do next week.

Also @candrews - it would be really nice if we could arrange a meeting to discuss how you and your corporate overlords are using HDF-Converters / the SAF.  We don't do any tracking, so we gotta go the old fashioned way and actually talk to our users in order to learn how/why they're using our stuff, and what we could do to deepen the partnership.

If you could email me at amann@mitre.org with some dates/times that you/your team/your management chain are available, I'd appreciate it a lot.

candrews commented:
> This pull request has a conflict. Could you fix it @candrews?

I've addressed the conflict.

candrews commented:
> Sorry for the delay @candrews - could you address the findings for the HDF-Converters tests? It seems like some of the other XML mappers such as the Veracode are now having issues.

As far as I can tell, all tests pass. If there is a problem, can you please point me to it?

candrews commented:
@Amndeep7 do you have an update as to when you'll be able to review and (hopefully) merge this request?

Amndeep7 commented on libs/hdf-converters/src/xccdf-results-mapper.ts:
```diff
@@ -278,142 +307,141 @@ export class XCCDFResultsMapper extends BaseConverter {
-                    _.pickBy(convertEncodedXmlIntoJson(description), _.identity)
-                  )
+                path: ['description.text', 'description'],
+                transformer: (description: string): string => {
```
you don't need a wrapping pair of braces + return on an arrow function if it's a single expression

candrews replied:
I've made that change (and the same change in other such places) :+1:

Amndeep7 commented on libs/hdf-converters/src/xccdf-results-mapper.ts:
```diff
@@ -278,142 +307,141 @@ export class XCCDFResultsMapper extends BaseConverter {
+              ident: {path: 'ident'},
+              reference: {path: 'reference'},
+              selected: {path: 'selected'},
+              version: {path: ['id', 'version']}, // dunno why the field is called version when it's just an old id
```
i think i had id in here as a bug, so we should be safe to make it only 'version' unless you've found id to also be relevant here.  i fixed it when i did the xml parser update.

candrews replied:
Removing `id` does change the output resulting in "version" not being set in some examples.

Perhaps we can consider that in a separate issue from this one?

Amndeep7 replied:
Sure we can make this a separate issue.  Maybe we change the order of attempted paths such that it attempts to find 'version' before using 'id' as a fallback?

candrews replied:
> Maybe we change the order of attempted paths such that it attempts to find 'version' before using 'id' as a fallback?

Sure, I've added a commit that does so.

Amndeep7 replied:
Looks great.  We can probably remove the comment there too :)

candrews replied:
I've added a commit that remove that comment :)

Amndeep7 commented on libs/hdf-converters/test/mappers/forward/xccdf_mapper.spec.ts:
```diff
@@ -54,6 +54,30 @@ describe('xccdf_mapper', () => {
+        )
+      );
+
+      // fs.writeFileSync(
```
ignore the sonarqube codesmell warning for the commented out code here.  if it has any other findings though, please do fix them.

candrews replied:
This is the only finding reported by sonarqube; I'll continue to ignore it :)

Amndeep7 commented on libs/hdf-converters/sample_jsons/xccdf_results_mapper/sample_input_report/xccdf-results-openscap-ComplianceAsCode-ubuntu1804.xml:
If you check out the deploy preview (https://deploy-preview-4255--heimdall-lite.netlify.app), it does not seem to parse + convert the XCCDF file in a reasonable amount of time.  The other files still completed within a few seconds.  This makes me feel like it's probably something to do with the recursion step.

If you see the results of running the tests on this pr, the xccdf unit tests took 155.569 s (https://github.com/mitre/heimdall2/actions/runs/4692756126/jobs/8318869062?pr=4255).  On some random other pr, it took 19.37 s (https://github.com/mitre/heimdall2/actions/runs/4921274476/jobs/8790946044).

candrews replied:
I found the culprit and fixed it: https://github.com/mitre/heimdall2/pull/4255/commits/c2640f507bf4862cd5496e24b01a79cf2f282200

The tests now run in 13 seconds; without this change, they took 53 seconds.

Amndeep7 commented on libs/hdf-converters/sample_jsons/xccdf_results_mapper/sample_input_report/xccdf-results-openscap-ComplianceAsCode-ubuntu1804.xml:
Please make sure to add/update the sample files in the saf cli as well.

candrews replied:
I'm assuming you mean at https://github.com/mitre/saf/tree/main/test/sample_data

That data would be updated as part of the PR that updates the version of the `@mitre/hdf-converters` dependency at https://github.com/mitre/saf/blob/73fa1def42f5eb39602d64debbe8a57a2ca5e2a2/package-lock.json#L223 which can only be done after this PR is merged and an updated version of this project (`@mitre/hdf-converters`) is released, right?

Amndeep7 replied:
you can just make a draft pr in that repo and say it's blocked by this one getting merged/released

candrews commented on libs/hdf-converters/src/xccdf-results-mapper.ts:
```diff
@@ -21,73 +21,56 @@ const IMPACT_MAPPING: Map<string, number> = new Map([
+  const matchingRuleResult = ruleResults.find(
+    (element) => _.get(element, 'idref') === ruleId
+  );
+  if (matchingRuleResult) {
```
Good point! I've made this change :+1:

candrews commented on libs/hdf-converters/src/xccdf-results-mapper.ts:
```diff
@@ -97,65 +80,63 @@ function convertEncodedXmlIntoJson(
   return parseXml(encodedXml);
```
I used to have it doing more... I didn't notice it became a useless wrapper :)

I've made the suggested change :+1:

candrews commented on libs/hdf-converters/src/xccdf-results-mapper.ts:
```diff
@@ -164,19 +145,14 @@ interface IIdent {
   const CCI_REGEX = /CCI-(\d*)/;
 
   const output: string[] = [];
   inputArray.forEach((element) => {
```
Sure, I've made this change :+1:

candrews commented on libs/hdf-converters/src/xccdf-results-mapper.ts:
```diff
@@ -21,73 +21,56 @@ const IMPACT_MAPPING: Map<string, number> = new Map([
+      return ExecJSON.ControlResultStatus.Skipped;
+    case 'notselected':
+      return ExecJSON.ControlResultStatus.Skipped;
+    case 'informational':
```
I've implemented this change :+1:

candrews commented:
@Amndeep7 I believe I've addressed all feedback and that all tests pass - can you please take another look?

candrews commented:
As far as I can tell, the ["Run E2E Backend + Frontend Tests / build (pull_request)"](https://github.com/mitre/heimdall2/actions/runs/4952309550/jobs/8859791454?pr=4255) test failure is unrelated to this PR - please correct me if I'm wrong.

Amndeep7 commented:
@candrews did you see my message on teams?  It seems like there are a handful of comments that you did not see/address yet.  Github's done this to me before where people haven't received some of the feedback that I wrote despite me nominally having submitted the review.  Do you see the below comment?

![image](https://github.com/mitre/heimdall2/assets/2015821/0a151473-7207-4426-a8ff-a0b7f55f2bec)

candrews commented on libs/hdf-converters/src/xccdf-results-mapper.ts:
```diff
@@ -164,19 +145,14 @@ interface IIdent {
   inputArray.forEach((element) => {
-    const text = _.get(element, 'text');
-    if (text.match(CCI_REGEX)) {
+    const text = String(_.get(element, 'text'));
```
I don't know `String` is there... I agree it shouldn't be there.

I've added a commit removing it.

candrews commented on libs/hdf-converters/src/xccdf-results-mapper.ts:
```diff
@@ -223,7 +252,7 @@ export class XCCDFResultsMapper extends BaseConverter {
           transformer: (input: Record<string, unknown>): string => {
             const descriptionPaths = [
-              ['description.text', 'description'],
+              ['description.text'],
```
Maybe? I'd have to study more xccdf files and see. However, this seems to be good for now, at least.

candrews commented on libs/hdf-converters/src/xccdf-results-mapper.ts:
```diff
@@ -223,7 +252,7 @@ export class XCCDFResultsMapper extends BaseConverter {
           transformer: (input: Record<string, unknown>): string => {
             const descriptionPaths = [
-              ['description.text', 'description'],
+              ['description.text'],
```
Because here, it's the path is used as an "and" not an "or"

The value used here is the combination of the list items (not a fallback, as is the case elsewhere). So if we used `description` and `description.text` here, then the value used would have the description twice, which is not desirable. That's observable by making this change then running the tests - the duplication is evident in the output.

candrews commented on libs/hdf-converters/src/xccdf-results-mapper.ts:
```diff
@@ -192,6 +168,59 @@ function nistTag(input: IIdent | IIdent[]): string[] {
+/**
+ * Given a group, returns all the rules within it (including rules in groups nested within the group).
+ */
+function getRulesInGroup(
```
I added tsdoc to this function to clarify how it works

candrews commented on libs/hdf-converters/src/xccdf-results-mapper.ts:
```diff
@@ -164,19 +145,14 @@ interface IIdent {
-    const text = _.get(element, 'text');
-    if (text.match(CCI_REGEX)) {
+    const text = String(_.get(element, 'text'));
+    if (!!text && text.match(CCI_REGEX)) {
```
Since the String conversion has been removed, the `!!text` part still serves a purpose (catching null / undefined)

candrews commented:
> did you see my message on teams?

Yes, I did

> It seems like there are a handful of comments that you did not see/address yet. Github's done this to me before where people haven't received some of the feedback that I wrote despite me nominally having submitted the review. Do you see the below comment? There are a few others as well.

I found https://github.com/mitre/heimdall2/pull/4255/files/88cd988c1d5771c7e5e0c5be16c010ae607cad6a..6992f6d7234f03ef14aab5d6d1ce027f981db659#r1187829883 and addressed it.

I also found and addressed these:
* https://github.com/mitre/heimdall2/pull/4255/files#r1187864828
* https://github.com/mitre/heimdall2/pull/4255/files#r1187868617
* https://github.com/mitre/heimdall2/pull/4255/files#r1187840749
* https://github.com/mitre/heimdall2/pull/4255/files#r1187830802

I _think_ that's everything... please let me know. Thank you!

Amndeep7 commented on libs/hdf-converters/src/xccdf-results-mapper.ts:
```diff
@@ -278,142 +299,149 @@ export class XCCDFResultsMapper extends BaseConverter {
               {
                 data: {
-                  path: 'Rule.check.check-content-ref.name',
+                  path: ['check.check-content-ref.name'],
```
<img width="1846" alt="image" src="https://github.com/mitre/heimdall2/assets/2015821/765799d1-06bd-4b01-baae-12648e07590f">

check text and references are falling back to the correspondingly named tags when they are left as empty string / empty array.  this causes [Object] to show up in heimdall when looking at the details tab.

please update descriptions.check and the refs array to contain this information instead.  refs array will probably be pretty easy - just take the data that looks like this (
![image](https://github.com/mitre/heimdall2/assets/2015821/26ccbd97-fe3c-4dc0-aabb-26663f4a1548)
) and put each one of those objects as that 'ref' attribute (https://github.com/mitre/heimdall2/blob/master/libs/inspecjs/src/generated_parsers/v_1_0/exec-json.ts#L245).  check text will be harder (
![image](https://github.com/mitre/heimdall2/assets/2015821/91bb9f8c-b6e1-4fbb-946c-67d344aceb81)
) - easiest thing to do that i'm fine with is just running a JSON.stringify and dumping it out.

Amndeep7 replied:
whereas the ref/reference stuff is not overlapped, the check subdescription and check tag do have overlap where if the subdescription is available it hides the tag.  for now, we have stringified the tag so it looks good in heimdall, but we should go and make the subdescription find things properly and behave well with the original data being an array for example.

Amndeep7 commented on libs/hdf-converters/src/xccdf-results-mapper.ts:
```diff
@@ -278,142 +299,149 @@ export class XCCDFResultsMapper extends BaseConverter {
-                    title: ['title.text', 'title']
-                  };
-                  return getProfiles(profiles, pathsSelect, paths);
+              ident: {path: 'ident'},
```
![image](https://github.com/mitre/heimdall2/assets/2015821/05088c26-64a7-41fb-a048-ed2ebbc621f0)

seems like there are more attributes that don't exist in the output coming from complianceascode.  we should probably toss those in the transformer as a conditionally provided attribute (see the version example).

candrews replied:
I believe I have now address this point.

Amndeep7 commented on libs/hdf-converters/src/xccdf-results-mapper.ts:
```diff
@@ -278,142 +299,149 @@ export class XCCDFResultsMapper extends BaseConverter {
               },
-              severity: {path: 'Rule.severity'},
+              severity: {path: 'severity'},
               description: {
```
![image](https://github.com/mitre/heimdall2/assets/2015821/dfd9948a-74bf-49d8-9086-0263f9b34905)

requires more work to handle the embedded html tags in there - "theso" came from "place a file in the <html blob> so that it" but the html blob and surrounding whitespace was removed for some reason

candrews replied:
This problem appears to be caused by shortcoming the XML parser used by this project: https://github.com/NaturalIntelligence/fast-xml-parser/issues/558

I don't know what, if anything, I can do about this one... any suggestions?

Amndeep7 replied:
Yeah, we rewrite it using JSONIX which hopefully just leaves random embedded blocks as they are.  So long as the with-raw option is provided, this isn't a blocker cause the user could at the end of the day retrieve this information.  I'm not gonna make this a blocker.

Amndeep7 commented on libs/hdf-converters/src/xccdf-results-mapper.ts:
```diff
@@ -278,142 +299,149 @@ export class XCCDFResultsMapper extends BaseConverter {
               nist: {
-                path: 'Rule.ident',
+                path: ['ident', 'reference'],
                 transformer: nistTag
```
not producing correct output at times.  see that same "crypto policies are means of" rule.  
![image](https://github.com/mitre/heimdall2/assets/2015821/a61fd3f2-24fd-4bf0-a4ab-24e6e6296168)
![image](https://github.com/mitre/heimdall2/assets/2015821/bf452ce6-2646-4f94-8f07-953e94d6a895)
the nist tag list should consist of those 5

Amndeep7 replied:
feel like it's cause we only look for cci's and then the corresponding nist tag.  if there are no cci's provided then we stuff the default nist tags in there (sa-11/ra-5) as a (usually correct but probably not this time) reasonable enough default.

however, if the references list provides actual nist tags as well then we should use those in addition to / instead of the cci ones.

another example of this happening is 'the installed operating system is fips 140-2 certified".  there are two cci's identified from the references list (
![image](https://github.com/mitre/heimdall2/assets/2015821/8a9dd3be-e89e-459c-ab29-27bdce141651)
) and so we get 2 nist tags, but the references section itself includes a hell of a lot more (
![image](https://github.com/mitre/heimdall2/assets/2015821/32665a99-b007-4fdd-9134-51de406c72f1)
)

candrews replied:
I believe I have now addressed this point.

Amndeep7 commented on libs/hdf-converters/src/xccdf-results-mapper.ts:
```diff
@@ -278,142 +299,149 @@ export class XCCDFResultsMapper extends BaseConverter {
               {
                 data: {
-                  path: 'Rule.fixtext.text',
+                  path: ['fixtext', 'fix'],
```
![image](https://github.com/mitre/heimdall2/assets/2015821/b1d64e02-83a2-4fd7-82cb-a6fd067027d7)
fix text not being found

Amndeep7 commented on libs/hdf-converters/src/xccdf-results-mapper.ts:
```diff
@@ -278,142 +299,149 @@ export class XCCDFResultsMapper extends BaseConverter {
+                  description
+                ) as string
             },
             descriptions: [
```
![image](https://github.com/mitre/heimdall2/assets/2015821/955c4aec-c771-4a6c-8a30-58a96eacd359)

rationale should probably be added as a conditional subdescription

candrews replied:
I believe I have now addressed this point.

Amndeep7 commented on libs/hdf-converters/src/xccdf-results-mapper.ts:
```diff
@@ -278,142 +299,149 @@ export class XCCDFResultsMapper extends BaseConverter {
+            path: 'Benchmark',
+            pathTransform: getRulesInBenchmark,
             key: 'id',
             tags: {
```
![image](https://github.com/mitre/heimdall2/assets/2015821/417cba1f-3c12-4376-ad50-01bf7e967613)

warning should be added as a conditionally supplied tag or subdescription

Amndeep7 commented on libs/hdf-converters/src/xccdf-results-mapper.ts:
```diff
@@ -278,142 +299,149 @@ export class XCCDFResultsMapper extends BaseConverter {
               },
-              group_id: {path: 'id'},
-              group_title: {path: ['title.text', 'title']},
+              group_id: {path: 'group.id'},
```
![image](https://github.com/mitre/heimdall2/assets/2015821/e352b6a6-b743-48b8-af45-69872335b5a6)
<img width="490" alt="image" src="https://github.com/mitre/heimdall2/assets/2015821/32750567-6345-437d-bf37-9aea3ea12c6e">

doubly nested groups only seem to show the immediate parent group's information.  i feel like all levels of group information should be shown here, but am not sure how best to reflect that information such that it shows up well within heimdall.  the also showcases other problems like how certain group level attributes don't get surfaced properly such as that 'value' block.  probably worthwhile setting up a brief call with you/me/aaron/eugene to figure out how to handle this one.

Amndeep7 replied:
making this a separate issue to figure out a good solution for multiply-nested groups.  for now, only showing the parent group is sufficient.

Amndeep7 commented on libs/hdf-converters/src/xccdf-results-mapper.ts:
```diff
@@ -278,142 +299,149 @@ export class XCCDFResultsMapper extends BaseConverter {
-                    return id && id === valueIdTracker;
-                  });
-                }
+                path: ['values']
```
this seems to get highly borked at times

for "configure auditd max_log_file_action upon reaching maximum log size"

![image](https://github.com/mitre/heimdall2/assets/2015821/e95eeccf-85d3-4bbe-9231-c2bdbe111017)

![image](https://github.com/mitre/heimdall2/assets/2015821/f152eebe-f916-47d9-9232-83e2794d09de)

Seems like it's extracting out the 'br' stuff from being embedded in the text content and putting it separate whereas the original document is using it as html-encoded formatting :(

candrews replied:
This problem appears to be caused by shortcoming the XML parser used by this project: https://github.com/NaturalIntelligence/fast-xml-parser/issues/558

I don't know what, if anything, I can do about this one... any suggestions?

Amndeep7 replied:
not considered a blocker - same as the other fast-xml-parser comment

Amndeep7 commented on libs/hdf-converters/sample_jsons/xccdf_results_mapper/xccdf-openscap-ComplianceAsCode-ubuntu1804-hdf.json:
```diff
@@ -32860,13 +32776,16 @@
                 "text": "SRG-OS-000059-VMM-000280",
                 "href": ""
               },
+              {
```
example of a validate duplicate showing up in the hdf file cause the original sample file has it duplicated as well for some reason

![image](https://github.com/mitre/heimdall2/assets/2015821/b6fef69e-2625-4d78-ae01-1462f9e0da97)

Amndeep7 commented on libs/hdf-converters/src/xccdf-results-mapper.ts:
```diff
@@ -278,142 +299,155 @@ export class XCCDFResultsMapper extends BaseConverter {
               {
                 data: {
-                  path: 'Rule.fixtext.text',
+                  path: ['fixtext', 'fix'],
```
![image](https://github.com/mitre/heimdall2/assets/2015821/1168f276-a24f-4b64-b85c-72ffe30d6b7e)

example of multiple fix texts existing that i don't think the code would be able to handle even if the path were correctly pointing to it

candrews commented on libs/hdf-converters/src/xccdf-results-mapper.ts:
```diff
@@ -392,8 +392,21 @@ export class XCCDFResultsMapper extends BaseConverter {
+                >;
+                if (ruleResult) {
+                  const result = _.get(ruleResult, 'result') as string;
+                  if (result == 'notapplicable' || result == 'informational') {
```
:+1: Done

candrews commented on libs/hdf-converters/src/xccdf-results-mapper.ts:
```diff
@@ -278,142 +299,155 @@ export class XCCDFResultsMapper extends BaseConverter {
                   transformer: getStatus
                 },
-                code_desc: '',
+                code_desc: {
```
This problem appears to be caused by shortcoming the XML parser used by this project: https://github.com/NaturalIntelligence/fast-xml-parser/issues/558

I don't know what, if anything, I can do about this one... any suggestions?

candrews commented:
@Amndeep7 I believe that I have addressed all the feedback.

As you know, I'm not terribly familiar with the heimdall tool, but with that disclaimer, I think the output looks good and is definitely improvement over when we last looked at this together.

Can you please take another look and let me know what else needs to be done?

Amndeep7 commented on libs/hdf-converters/src/xccdf-results-mapper.ts:
```diff
@@ -375,7 +421,19 @@ export class XCCDFResultsMapper extends BaseConverter {
               })
             },
-            refs: [],
+            refs: [
```
from xccdf-scc-rhel8-hdf.json

![image](https://github.com/mitre/heimdall2/assets/2015821/b6f6d2b1-6654-4d1b-80f5-559e6f4d1829)

your sample outputs don't match the schema for some reason

https://github.com/mitre/heimdall2/blob/master/libs/inspecjs/src/generated_parsers/v_1_0/exec-json.ts#L244

should probably look something like `refs: [ref: [{publisher: blah, ...}]]`

candrews replied:
That data structure is a bit confusing... but I think I matched the intent now.

Amndeep7 commented on libs/hdf-converters/src/xccdf-results-mapper.ts:
```diff
@@ -392,16 +450,34 @@ export class XCCDFResultsMapper extends BaseConverter {
-                  transformer: parseHtml
+                  path: ['fixtext.text', 'fixtext', 'fix.text', 'fix'],
+                  transformer: (text: string | string[]) =>
+                    asArray(text).map(parseHtml).join('\n') || undefined
```
the '|| undefined' is dangerous cause it makes it not match the schema

xccdf-scc-rhel8-hdf.json

![image](https://github.com/mitre/heimdall2/assets/2015821/ffcc8c78-91be-43d8-bb72-eb1e2e4a025f)

schema: https://github.com/mitre/heimdall2/blob/master/libs/inspecjs/src/generated_parsers/v_1_0/exec-json.ts#L224

if a description is provided it needs to have both a label and data.  the or defined makes it so that sometimes only a label is provided.

candrews replied:
Updated! :+1:

Amndeep7 commented on libs/hdf-converters/src/xccdf-results-mapper.ts:
```diff
@@ -151,11 +160,18 @@ function extractCci(input: IIdent | IIdent[]): string[] {
+      .map((x) => x.text)
+      .map(parse_nist)
+      .filter((x) => !!x)
+      .filter(is_control)
```
seems like your sample is not updated properly still?  or this function is borked lol

xccdf-scc-rhel7-hdf.json

![image](https://github.com/mitre/heimdall2/assets/2015821/6345d9a5-84fc-4eef-8a1a-c80120c776ff)

seems like the nist tag array is including the stig control ids, which it shouldn't do

candrews replied:
`is_control(parse_nist('SV-86515'))` return true... so as far as inspectjs is concerned, 'SV-86515' is a nist control and therefore should be included.

Am I misunderstanding `nist.js`? How should I determine if an item should be included or excluded from this list?

Amndeep7 replied:
this feels like a failure on inspecjs' part in not removing things that are clearly wrong. parse_nist should throw errors/explicitly fail when they are not nist controls/control families. it not doing so confuses the type system by mistakenly typing that string. is_control more or less is only used to narrow down types and doesn't do much actual validation.

anyways as shown there and in 'raw_nist.ts', nist controls look like 'two_letter_family-small_number possibly followed by letters and numbers in parentheses' whereas stig ids are 'S?V-5__or_6_digit_number'.

the issue that arose here is that it thinks 'sv' is a control family and then processes the numbers after it as like 'control "big number" from within that control family' but doesn't take into account the reality of the situation where 'sv' is not a valid control family nor is there a control with an index as high as '86515' in existence.

looking briefly, it seems like the regexes that parse_nist uses are vastly broader than the actually allowed set of inputs. i will experiment to see if it's possible to restrict it to allowed control families.

Amndeep7 commented on libs/hdf-converters/src/xccdf-results-mapper.ts:
none of the original or hdf samples open up in heimdall i think cause of the schema issues

attempt at raw file:  
![image](https://github.com/mitre/heimdall2/assets/2015821/7ef0df17-ef2b-4b39-8753-e8651938007f)

attempt with generated hdf file:  
![image](https://github.com/mitre/heimdall2/assets/2015821/b10a55ad-5668-41f7-97e9-74e05d76ddff)

this is it complaining about the controldescription type not being matched:
![image](https://github.com/mitre/heimdall2/assets/2015821/c13e0a97-1990-426a-a6d4-6eb4bc9f06fb)

candrews replied:
I tested a bunch and they load now.

If you have one that doesn't, please give me the name so I can take a look.

Amndeep7 reviewed (changes requested):
Feeling like this should probably be able to be merged in after this final set of changes is made.  Thanks for bearing with the process for so long.

candrews commented:
Using `stopNodes` would be a great way to fix the parsing problem that occurs with nodes that contain HTML.

However... `stopNodes` doesn't work for XML documents that use namespaces: https://github.com/NaturalIntelligence/fast-xml-parser/blob/v4.2.4/src/xmlparser/OrderedObjParser.js#L308 Attempting to use it results in an "Unexpected end of <tag>" error.

If this issue in the xml parser is fixed, then modifying `xccdf-results-mapper.ts` to say:
```javascript
  constructor(scapXml: string, withRaw = false) {
    super(
      parseXml(scapXml, {
        stopNodes: [
          '*.fixtext',
          '*.fix',
          '*.rationale',
          '*.warning',
          '*.title',
          '*.description'
        ]
      })
    );
```

would solve the html parsing problem.

Amndeep7 commented on libs/hdf-converters/src/xccdf-results-mapper.ts:
```diff
@@ -481,37 +481,41 @@ export class XCCDFResultsMapper extends BaseConverter {
+                  data: asArray(data).join('\n'),
+                  label: 'check'
+                })
+              } as any,
```
would prefer to avoid using 'any' if possible?  how does it complain if you don't cast it?

if we can't make it happy by clearing up the types directly, then i usually prefer doing 'as unknown as whatever_it_ought_to_be' over 'as any'

candrews replied:
The type of `descriptions` is `ControlDescription` and that has two mandatory fields: `data` and `label`. Neither of those fields are directly specified (they are set by the transformer, which typescript has no understanding of).

I implemented the `as unknown as ControlDescription` approach.

Amndeep7 commented on libs/inspecjs/src/nist.ts:
```diff
@@ -2,9 +2,10 @@ import {ControlStatus} from './compat_wrappers';
 
 // Regexes.
-const NIST_FAMILY_RE = /^[A-Z]{2}$/;
+// Matches against only currently existing NIST Control families instead of allowing any two adjacent uppercase letters
```
<img width="601" alt="image" src="https://github.com/mitre/heimdall2/assets/2015821/a1c8ac5f-efee-4c3c-a1ef-00442b0a5934">

seems like it is no longer falsely validating sv numbers as controls but

<img width="697" alt="image" src="https://github.com/mitre/heimdall2/assets/2015821/71e08322-ccd6-4743-9973-1dfa59a49eef">

these undefineds are getting passed through which ain't good cause they're not valid json.  idk where they're coming from tho - don't *think* it should have anything to do with the change that i made

candrews replied:
I don't see `undefined` in any of the `libs/hdf-converters/sample_jsons/xccdf_results_mapper/*.json` files. Can you link me to where I can see it or provide the file name that I can look at to see this problem?

Amndeep7 replied:
I got these by doing the yarn run test.  you can see them in the logs on github as well.

candrews replied:
I don't see this in any of the github logs, nor in the output of `yarn test`, nor or any of the files.

Is it possible you're looking at an old build or old code?

If you can find it, can you please link me to it, or give me the path of a file and line number to took at?

Amndeep7 replied:
Realized what was happening: the test is comparing the output of the file (JSON) with the output from running the toHdf function (JS).  consequently it sees the 'attribute: unknown' in there cause that's still valid in js (and is functionally a no-op) but it gets removed when turned into json.  when you get rid of any other differences in the files (ex. the differently formatted / duplicate nist tags), then it seems like jest is smart enough to realize that they are equal and doesn't surface anything into the console.

Amndeep7 commented on libs/inspecjs/src/nist.ts:
```diff
@@ -2,9 +2,10 @@ import {ControlStatus} from './compat_wrappers';
 // Regexes.
-const NIST_FAMILY_RE = /^[A-Z]{2}$/;
+// Matches against only currently existing NIST Control families instead of allowing any two adjacent uppercase letters
+const NIST_FAMILY_RE = /^(?:A[CPRTU]|C[AMP]|D[IM]|I[APR]|M[AP]|P[ELMS]|RA|S[ACEI]|TR|U[LM])$/;
```
would appreciate you pleasing the linter on my behalf and also regenerating the samples

candrews replied:
Done! The linting issues have been addressed and samples regenerated.

Amndeep7 commented on libs/hdf-converters/src/xccdf-results-mapper.ts:
```diff
@@ -164,34 +145,93 @@ interface IIdent {
   return output;
 }
 
 function nistTag(input: IIdent | IIdent[]): string[] {
```
<img width="689" alt="image" src="https://github.com/mitre/heimdall2/assets/2015821/d291373d-391c-413a-9909-dbc0fcbefbb7">

I think the function in that utility file is expecting the nist tag to only be ex. 'SI-2' whereas we are passing in ex. 'SI-2 (b)'.  Considering that the "800-53 Controls & CCIs" column in the table can show the full thing, I think this is a bug with the utility file where it's not dealing with the full string properly.  If you have the time to debug it, I'd appreciate it, otherwise I'll get to it myself in the next couple days.

candrews replied:
Have you had a chance to address this issue?

Amndeep7 replied:
Turns out that it was not a problem with the specification but with the zero padding that was causing issues.  Anyways, I've made the fix.  I will merge if there are no other issues that arise during the CI.

candrews commented:
@Amndeep7 do you have a guess as to when we'll be able to progress this PR?

Amndeep7 commented:
Sorry I've been swamped this week.  I'll try to take a look next week to fix that bug in Heimdall that I mentioned.

Amndeep7 commented:
Thank you for your patience and hard work on this PR @candrews.  You did some absolutely fantastic work here that'll be appreciated across many of the users of Heimdall.
